### ISSUE TITLE
Wrong warning for multiple type parameters if argument passed is a variable

### ISSUE THREAD

**Issue by pksadiq** [python/mypy#18305]
**Bug Report**
When the function parameter is annotated to have some combination of multiple types with `|`, mypy validation warns
if a variable is passed to the called function.

**To Reproduce**

```python
from typing import Any

def show(data: dict[int | str, dict[str, str]]):
    print(data)

data = {0: {"message": "Hello world"}}

# Warning
show(data)

# No warning
show({0: {"message": "Hello world"}})
```
Gist url: https://mypy-play.net/?mypy=latest&python=3.12&gist=2e3c8b121347ac9547e8e84eba7f1017

**Expected Behavior**

There should be no warning as it's a valid code

**Actual Behavior**

```sh
mypy ./test.py 
test.py:10: error: Argument 1 to "show" has incompatible type "dict[int, dict[str, str]]"; expected "dict[int | str, dict[str, str]]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used:  1.13.0 (compiled: yes) as packaged in Debian sid
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.12.8

**Comment by brianschubert** [comment]
Hi! This is https://mypy.readthedocs.io/en/stable/common_issues.html#invariance-vs-covariance

Like many mutable generic containers, `dict` is [invariant](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generics) with respect to its type parameters. That means that even though `int` is a subtype of `int | str`, `dict[int, V]` is not a subtype of `dict[int | str, V]`. To see why that would be unsafe, consider this example:
```python
def foo(x: dict[int | str, str]):
    x["a"] = "bad"

d: dict[int, str] = {1: "good"}
foo(d)

for key in d:
    assert isinstance(key, int)  # boom!
```

The reason the second example works is because bidirectional inference allows mypy to infer a better type for the dict literal by peeking at the type of the parameter that its being passed to.

**Comment by pksadiq** [comment]
I see. I can't modify the function signature. I shall annotate my variable instead.

Thanks for the explanation.
